tests/integrationv2/test_serialization.py

```diff
@@ -0,0 +1,93 @@
+import pytest
+import copy
+import os
+import random
+import string
+
+from configuration import available_ports
+from common import ProviderOptions, Protocols
+from fixtures import managed_process  # lgtm [py/unused-import]
+from providers import Provider, S2N
+from utils import invalid_test_parameters, get_parameter_name, to_bytes
+
+SERVER_STATE_FILE = 'server_state'
+CLIENT_STATE_FILE = 'client_state'
+
+SERVER_DATA = f"Some random data from the server:" + "".join(
+    random.choice(string.ascii_uppercase + string.digits) for _ in range(10)
+)
+CLIENT_DATA = f"Some random data from the client:" + "".join(
+    random.choice(string.ascii_uppercase + string.digits) for _ in range(10)
+)
+
+"""
+This test file checks that a serialized connection can be deserialized by an older version of
+s2n-tls and vice versa. This ensures that any future changes we make to the handshake are backwards-compatible
+with an older version of s2n-tls.
+
+This feature requires an uninterrupted TCP connection with the peer in-between serialization and
+deserialization. Our integration test setup can't provide that while also using two different s2n-tls
+versions. To get around that we do a hack and serialize/deserialize both peers in the TLS connection.
+This prevents one peer from receiving a TCP FIN message and shutting the connection down early.
+"""
+
+
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13, Protocols.TLS12], ids=get_parameter_name)
+@pytest.mark.parametrize("use_mainline_version", [True, False], ids=get_parameter_name)
```

Comment:
https://github.com/aws/s2n-tls/pull/4548/files#r1600387435 I think Sam's confusion here is because of the naming. You use the mainline version regardless of True/False here. It's more about which you use to serialize vs deserialize. So maybe "mainline_serializes"? Or "use_mainline_to_serialize"? But even that doesn't really capture that if mainline doesn't serilaize, then it deserializes.

Maybe an enum with values that are more readable than True/False? Like MainlineRole.Serialize, MainlineRole.Deserialize?

Reply:
I think the enums do make this test more readable? lmk what you think!